src/linker/Linker.Dataflow/MethodBodyScanner.cs

```diff
@@ -709,7 +709,9 @@ private static void ScanLdtoken (Instruction operation, Stack<StackSlot> current
 			}
 
 			if (operation.Operand is TypeReference typeReference) {
-				var resolvedReference = typeReference.Resolve ();
+				var resolvedReference = typeReference is ArrayType ?
+					typeReference.Module.ImportReference (typeof (Array))?.Resolve () : typeReference.Resolve ();
```

Comment:
Can you please introduce a helper method for this? Maybe an extension method on `TypeReference` - something like `ResolveToMainTypeDefinition` ... I can't come up with a better name - maybe @MichalStrehovsky  will know. It should:
* For normal types it does what `.Resolve` does today
* For generics it resolves to the open generic type (which is what `.Resolve` does today as well)
* For arrays it resolves to `System.Array` type def.
* The question is what should happen for other "weird" cases like - multi-dim arrays, pointers, by-refs, ... we should probably look into this some more if it can every happen and what does it mean when it does happen.

Eventually (We should either fix this right away or file an issue) this should handle not just arrays, but pointers, by-refs and so on.

Reply:
I added this as an extension method as you suggested. Note that for multidim arrays it will behave the same way as for unidimensional arrays, since Cecil does not differentiate between the two (both are `ArrayType`). Not sure if we want different "resolve semantics" for pointers or by-refs though.